### ISSUE TITLE
AllenNLP retreived answer span confidence score

### ISSUE THREAD

**Issue by shivamani-ans** [allenai/allennlp#1265]
Hi,
I am exploring on AllenNLP. Can I also get confidence score of retrieved answer span from paragraph.

 Thank you.

**Comment by murphp15** [comment]
The way the SQUAD models are trained they assume that the answer is always in a accompanying paragraph. 
Because of this reason it doesn't really generate a simple confidence score that can be used to determine how confident if is of the answer being correct.
I hope that answers your question?

**Comment by shivamani-ans** [comment]
Thank you for your response.

But actually i have multiple paragraphs for single question, so i would like to check confidence score based on score i would like to confirm the answer.

for example: i have three paragraphs below for same question from a document, how i can judge correct answer.

{"passage": "Record the time or days taken for the evaporation process in the above cases.", "question": "What is evaporation?"}

{"passage": "We know that evaporation is a surface phenomenon. If the surface area is increased, the rate of evaporation increases. For example, while putting clothes for drying up we spread them out.", "question": "What is evaporation?"}

{"passage": "We know that particles of matter are always moving and are never at rest. At a given temperature in any gas, liquid or solid, there are particles with different amounts of kinetic energy. In the case of liquids, a small fraction of particles at the surface, having higher kinetic energy, is able to break away from the forces of attraction of other particles and gets converted into vapour. This phenomenon of change of a liquid into vapours at any temperature below its boilingpoint is called evaporation.", "question": "What is evaporation?"}

Thank you.

**Comment by murphp15** [comment]
@shivamani-ans  
If the answer is definitely in the one of the paragraphs then you can use the concept as described in https://arxiv.org/pdf/1710.10723.pdf. 
The model for that paper is in this PR https://github.com/allenai/allennlp/pull/804. 
However this model is still trained under the assumption that the answer exists in one of the paragraphs. 
What is your business use case, I think I am probably working on something quite similar to what you are trying to do?

**Comment by shivamani-ans** [comment]
Yeah, I have a medical document with about 100 paragraphs. I would like to retrieve a right paragraph answering the question i ask.

**Comment by murphp15** [comment]
I am trying to do something very similar. 
Presumably if a client asks a question and the answer is not available in the document then it is not acceptable to return the wrong answer? 
Instead you need to return no answer.

**Comment by shivamani-ans** [comment]
The problem here is, there is an answer in the document but in Paragraph No 3, there are relevant words in 1st and 2nd paragraph. By default I am getting 1st paragraph as answer. When I remove 1st and 2nd paragraph from document, only then I get the 3rd paragraph. What I am hoping is to get all three paragraphs as answers but with "Confidence Score", So that I can pick the paragraph with HIGH Confidence.

**Comment by murphp15** [comment]
The PR I mentioned above is what you are looking for. 
But you will always be returning an answer to the client even if the answer is not in the document? 
In this case you are returning incorrect answers?

**Comment by shivamani-ans** [comment]
Hey, I am sorry, May be I am not explaining my problem in the right way :) .

I totally agree that if we dont get the right answer , I should be not be showing the answer to client.

But what I am looking for, is a way to programatically find if a Paragraph is correct or not depending on its "Confidence Score"

**Comment by murphp15** [comment]
@shivamani-ans 
Yeah i thought so, I am working on the exact same problem. 
Would you be interested in working on this together?

**Comment by murphp15** [comment]
@shivamani-ans my email address is murphp15@gmail.com add me and we can chat in google hangouts

**Comment by shivamani-ans** [comment]
Thank you, sent you an invite in google hangouts

**Comment by murphp15** [comment]
@schmmd I see you have added this to the backlog does this mean that you guys might create a module which meets the requirements as specified in this ticket?

**Comment by schmmd** [comment]
@murphp15 that was actually a mistake!  Sorry about that--we have no specific plans to address this and it's still unassigned.

**Comment by murphp15** [comment]
:-(

**Comment by matt-gardner** [comment]
Sorry to be slow answering the original question - it seemed like you two had figured out the issue.

So, to answer the first question, BiDAF already returns a confidence score.  The logits that you see here are the confidence scores: https://github.com/allenai/allennlp/blob/b26031db4f27d606146f7d5820b1e884559bd4d4/allennlp/models/reading_comprehension/bidaf.py#L252-L259

@murphp15 already answered the broader question, though, that these confidence scores will not necessarily be very well calibrated, and #804 will add a model that addresses this.

I don't think there are any other questions or issues here, so I'm going to close this issue.  Feel free to re-open it if I missed something or if you have further questions.

**Comment by murphp15** [comment]
Ok cool!
If you were building a production ready system like this would you decide on the logit score through simple observational analysis?

**Comment by matt-gardner** [comment]
I would construct a validation set from the data that you care about and measure the performance on that set as you vary this threshold, then pick the value of the threshold that best trades off precision and recall for your use case.

**Comment by rajurajvijay619** [comment]
Given n passages and a query, can we get the confidence scores of how correctly a passage answers a query?
Ex. Query - Q.
Candidate passages - p1,p2,p3,.....pn.
Rank them in order of relevancy/confidence scores(how correctly they answer the score)
p4
p2
p8
p1
.......
Is there any similar provision in allennlp, or anywhere? Thank you.

**Comment by matt-gardner** [comment]
The output of `model.forward()` (or `predictor.predict()`) returns the logits for each span, which you can use to recover what you're looking for.  There is nothing built-in that does exactly what you want, but it shouldn't be hard to do it with what's there.

**Comment by rajurajvijay619** [comment]
Thank you @matt-gardner .

**Comment by Ravibth** [comment]
Hi 
Allennlp always return the value. How to check the result score in the allennlp. 
![image](https://user-images.githubusercontent.com/13044714/68118200-641ce980-ff25-11e9-85dc-ee2f23fd5dda.png)
How to verified the our result?

**Comment by qw1243** [comment]
Has anyone got the solution to this problem.?I need to display confidence score of allennlp bidaf model for a QA system project I've made.

**Comment by anoop-pm** [comment]
howto find confidence score

**Comment by anoop-pm** [comment]
How to find the confident score to find Very Confident, Somewhat Confident
